### PR TITLE
Plumb through EarlyEOF to callers

### PR DESCRIPTION
**Pull request by kevinburke**
If we receive an EarlyEOF, send this through to the calling program,
so we can hit its callback and terminate it, instead of hanging
forever.

Verified this patch works by adding it to the Sourcegraph node_modules
directory, and sending Ctrl+C to the server - it terminates the
running process with an error instead of hanging forever.

### CODE REVIEW

File: src/WorkerPool.js [hunk 1]
```diff
@@ -94,6 +112,19 @@ class PoolWorker {
   onWorkerMessage(message, finalCallback) {
     const { type, id } = message;
     switch (type) {
+      case 'end': {
+        var err = new Error('worker terminated');
```
Comment (felixfbecker) [inline src/WorkerPool.js]: Use const instead of var

File: src/WorkerPool.js [hunk 2]
```diff
@@ -94,6 +112,19 @@ class PoolWorker {
...
+        for (var id in this.jobs) {
```
Comment (felixfbecker) [inline src/WorkerPool.js]: This also iterates inherited jobs, use Object.keys()

File: src/readBuffer.js [hunk 3]
```diff
@@ -55,7 +55,17 @@ export default function readBuffer(pipe, length, callback) {
...
     pipe.on('end', onEnd);
+=======
+    pipe.on('end', function() {
+      if (terminated) {
+        return;
+      }
+      terminated = true;
+      callback(new EarlyEOF('stream ended ' + remainingLength.toString() + ' bytes prematurely'));
+    })
+>>>>>>> Listen for end events
```
Comment (felixfbecker) [inline src/readBuffer.js]: Bad merge

File: src/WorkerPool.js (hunk 1 above)
Comment (kevinburke) [reply src/WorkerPool.js]: Done

File: src/WorkerPool.js (hunk 2 above)
Comment (kevinburke) [reply src/WorkerPool.js]: Done

File: src/readBuffer.js (hunk 3 above)
Comment (kevinburke) [reply src/readBuffer.js]: Fixed, sorry!

File: src/WorkerPool.js [hunk 4]
```diff
@@ -94,6 +112,21 @@ class PoolWorker {
   onWorkerMessage(message, finalCallback) {
     const { type, id } = message;
     switch (type) {
+      case 'end': {
+        const err = new Error('worker terminated');
+        const jobIds = Object.keys(this.jobs);
+        for (let i = 0; i < jobIds.length; i++) {
```
Comment (felixfbecker) [inline src/WorkerPool.js]: You can use `for (const key of Object.keys(this.jobs)`

Comment (kevinburke) [reply src/WorkerPool.js]: Done!

**Review by felixfbecker — APPROVED** [review]

**Pull request merged**